sklearn/tree/_tree.pyx

```diff
@@ -1132,3 +1132,284 @@ cdef class Tree:
+            # less than or equal to self.ccp_alpha
+            if cur_ccp_alpha > ccp_alpha:
+                break
+
+            # stack uses only the start variable
+            rc = stack.push(cur_idx, 0, 0, 0, 0, 0, 0)
+            if rc == -1:
+                with gil:
+                    raise MemoryError()
+
+            # descendants of branch are not in subtree
+            while not stack.is_empty():
+                stack.pop(&stack_record)
+                node_idx = stack_record.start
+                candidate_nodes[node_idx] = 0
+                leaves_in_subtree[node_idx] = 0
+                in_subtree[node_idx] = 0
+
+                child_l_idx = child_l[node_idx]
+                child_r_idx = child_r[node_idx]
+                if child_l_idx != child_r_idx:
+                    stack.push(child_l_idx, 0, 0, 0, 0, 0, 0)
+                    stack.push(child_r_idx, 0, 0, 0, 0, 0, 0)
+            leaves_in_subtree[cur_idx] = 1
+            in_subtree[cur_idx] = 1
+
+            # updates number of leaves
+            n_pruned_leaves = n_leaves[cur_idx] - 1
+            n_leaves[cur_idx] = 0
+
+            # # computes the increase in r_branch to bubble up
+            r_diff = r_node[cur_idx] - r_branch[cur_idx]
+            r_branch[cur_idx] = r_node[cur_idx]
+
+            # bubble up values to ancestors
+            cur_idx = parent[cur_idx]
+            while cur_idx != -1:
+                n_leaves[cur_idx] -= n_pruned_leaves
+                r_branch[cur_idx] += r_diff
+                cur_idx = parent[cur_idx]
+
+        for i in range(in_subtree.shape[0]):
+            if in_subtree[i]:
+                init_capacity += 1
+
+    _build_pruned_tree(tree, orig_tree, leaves_in_subtree, init_capacity)
+
+
+cdef _build_pruned_tree(
+    Tree tree, # OUT
+    Tree orig_tree,
+    const bint[:] leaves_in_subtree,
+    SIZE_t initial_capacity):
+    """Builds a pruned tree.
+
+    Builds a pruned tree from the original tree. The values and nodes from the
+    original tree are copied into the pruned tree.
+
+    Parameters
+    ----------
+    tree : Tree
+        Location to place the pruned tree
+    orig_tree : Tree
+        Original tree
+    leaves_in_subtree : bint memoryview
+        Boolean mask for leaves to include in subtree
+    init_capacity : SIZE_t
+        Number of nodes to initially allocate in pruned tree
+    """
+    tree._resize(initial_capacity)
+
+    cdef:
+        SIZE_t orig_node_id
+        SIZE_t new_node_id
+        SIZE_t depth
+        SIZE_t parent
+        bint is_left
+        bint is_leaf
+
+        # value_stride for original tree and new tree are the same
+        SIZE_t value_stride = orig_tree.value_stride
+        SIZE_t max_depth_seen = -1
+        int rc = 0
+        Node* node
+        double* orig_value_ptr
+        double* new_value_ptr
+
+        # Only uses the start, depth, parent, and is_left variables
+        Stack stack = Stack(INITIAL_STACK_SIZE)
```

Comment:
- To answer your previous comment, I think it's fine to use the pre-existing stack class since you make it clear that you only use a subset of the attributes
- just for my own understanding, you don't actually need a stack right? that could be done with a queue or anything else? (just to be clear I'm not saying you should change this I just want to make sure that `tree._add_node` doesn't except a DFS order ;))